Quarkus Quartz: Quarkus 1.13.x scheduler is not updated in DB when updated after upgrading to Quarkus 2.x

### Describe the bug

Hi everyone!
Some months ago, some schedulers were created using Quarkus 1.13.x and were deployed to production. Everything worked fine. Some weeks ago, we had the need to update the cron job of one of the scheduler but found out that the scheduler was updated in local tests but not in production. Currently we're running Quarkus 2.5.1.

After some digging in Quartz scheduler code, we found out that Quartz is not able to find the trigger because trigger name has changed from Quarkus 1.13.x to 2.x. 

The query performed by Quartz is (from org.quartz.impl.jdbcjobstore.StdJDBCDelegate#selectTrigger)
`"SELECT * FROM QUARTZ_TRIGGERS WHERE SCHED_NAME = 'QuarkusQuartzScheduler' AND TRIGGER_NAME = ? AND TRIGGER_GROUP = ?"`

Below one can see that the column `trigger_name` from `QUARTZ_TRIGGERS` table has different values when the schedulers are created in Quarkus 1.13.x and 2.5.1: _trigger was added to the identity of the scheduler 

Quarkus 1.13.7
| sched_name             | trigger_name          | trigger_group                  | job_name              |
|------------------------|-----------------------|--------------------------------|-----------------------|
| QuarkusQuartzScheduler | executeDataRemoverJob_trigger | io.quarkus.scheduler.Scheduler | executeDataRemoverJob |
| QuarkusQuartzScheduler | executeCalculationJob_trigger | io.quarkus.scheduler.Scheduler | executeCalculationJob |

Quarkus 2.5.1
| sched_name             | trigger_name          | trigger_group                  | job_name              |
|------------------------|-----------------------|--------------------------------|-----------------------|
| QuarkusQuartzScheduler | executeDataRemoverJob | io.quarkus.scheduler.Scheduler | executeDataRemoverJob |
| QuarkusQuartzScheduler | executeCalculationJob | io.quarkus.scheduler.Scheduler | executeCalculationJob |

When doing the query to find the trigger, none is found and no update is done. Since the job_name is the same, no new scheduler is created. Also we deleted one of the scheduler from the code but the trigger is still in the table. I think it's not deleted due to the same problem.

I checked the migration guide to 2.x and I didn't see any mention about this changes in Quartz.
Can you tell me if this is a known issue?

I didn't yet try any workaround but I will try some solution and post it here. 

### Expected behavior

Schedulers created with Quarkus 1.13.x should be updated when the update is done running the app with Quarkus 2.x.

### Actual behavior

Schedulers are not updated

### How to Reproduce?

1. create a Scheduler using Quarkus 1.13.x
2. upgrade the app to Quarkus 2.5.1
3. update the scheduler cron in the code
4. start the app: scheduler still runs using the old cron (DB is not updated)

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk 11.0.11 2021-04-20 OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9) OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.5.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

I changed the `identity` of the scheduler from `executeCalculationJob` to `executeCalculationJob_trigger` as suggested but quarkus doesn't start.

```
2021-12-23 08:54:30.509 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main)  Failed to start quarkus: java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.dev.appstate.ApplicationStateNotification.waitForApplicationStart(ApplicationStateNotification.java:51)
	at io.quarkus.runner.bootstrap.StartupActionImpl.runMainClass(StartupActionImpl.java:122)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:145)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:450)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:67)
	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:149)
	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:105)
	at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:145)
	at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:63)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:104)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:67)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:120)
	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:103)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalStateException: Unable to create Scheduler
	at io.quarkus.quartz.runtime.QuartzScheduler.<init>(QuartzScheduler.java:221)
	at io.quarkus.quartz.runtime.QuartzScheduler_Subclass.<init>(Unknown Source)
	at io.quarkus.quartz.runtime.QuartzScheduler_Bean.create(Unknown Source)
	at io.quarkus.quartz.runtime.QuartzScheduler_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
	at io.quarkus.quartz.runtime.QuartzScheduler_Bean.get(Unknown Source)
	at io.quarkus.quartz.runtime.QuartzScheduler_Bean.get(Unknown Source)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:440)
	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:453)
	at io.quarkus.arc.impl.ArcContainerImpl.instanceHandle(ArcContainerImpl.java:424)
	at io.quarkus.arc.impl.ArcContainerImpl.instance(ArcContainerImpl.java:233)
	at io.quarkus.arc.runtime.BeanLookupSupplier.get(BeanLookupSupplier.java:29)
	at io.quarkus.vertx.http.runtime.devmode.DevConsoleRecorder.addInfo(DevConsoleRecorder.java:41)
	at io.quarkus.deployment.steps.DevConsoleProcessor$runtimeTemplates-308161071.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.DevConsoleProcessor$runtimeTemplates-308161071.deploy(Unknown Source)
	... 13 more
Caused by: org.quartz.ObjectAlreadyExistsException: Unable to store Trigger with name: 'executeCalculationJob_trigger' and group: 'io.quarkus.scheduler.Scheduler', because one already exists with this identification.
	at org.quartz.impl.jdbcjobstore.JobStoreSupport.storeTrigger(JobStoreSupport.java:1184)
	at org.quartz.impl.jdbcjobstore.JobStoreSupport$2.executeVoid(JobStoreSupport.java:1068)
	at org.quartz.impl.jdbcjobstore.JobStoreSupport$VoidTransactionCallback.execute(JobStoreSupport.java:3780)
	at org.quartz.impl.jdbcjobstore.JobStoreSupport$VoidTransactionCallback.execute(JobStoreSupport.java:3778)
	at org.quartz.impl.jdbcjobstore.JobStoreCMT.executeInLock(JobStoreCMT.java:245)
	at org.quartz.impl.jdbcjobstore.JobStoreSupport.storeJobAndTrigger(JobStoreSupport.java:1063)
	at org.quartz.core.QuartzScheduler.scheduleJob(QuartzScheduler.java:855)
	at org.quartz.impl.StdScheduler.scheduleJob(StdScheduler.java:249)
	at io.quarkus.quartz.runtime.QuartzScheduler.<init>(QuartzScheduler.java:200)
	... 32 more
```

> I changed the `identity` of the scheduler from `executeCalculationJob` to `executeCalculationJob_trigger` as suggested but quarkus doesn't start.
> 
> ```
> 2021-12-23 08:54:30.509 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main)  Failed to start quarkus: java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
> 	at io.quarkus.dev.appstate.ApplicationStateNotification.waitForApplicationStart(ApplicationStateNotification.java:51)
> 	at io.quarkus.runner.bootstrap.StartupActionImpl.runMainClass(StartupActionImpl.java:122)
> 	at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:145)
> 	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:450)
> 	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:67)
> 	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:149)
> 	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:105)
> 	at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:145)
> 	at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:63)
> Caused by: java.lang.RuntimeException: Failed to start quarkus
> 	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
> 	at io.quarkus.runtime.Application.start(Application.java:101)
> 	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:104)
> 	at io.quarkus.runtime.Quarkus.run(Quarkus.java:67)
> 	at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
> 	at io.quarkus.runtime.Quarkus.run(Quarkus.java:120)
> 	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 	at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:103)
> 	at java.base/java.lang.Thread.run(Thread.java:829)
> Caused by: java.lang.IllegalStateException: Unable to create Scheduler
> 	at io.quarkus.quartz.runtime.QuartzScheduler.<init>(QuartzScheduler.java:221)
> 	at io.quarkus.quartz.runtime.QuartzScheduler_Subclass.<init>(Unknown Source)
> 	at io.quarkus.quartz.runtime.QuartzScheduler_Bean.create(Unknown Source)
> 	at io.quarkus.quartz.runtime.QuartzScheduler_Bean.create(Unknown Source)
> 	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
> 	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
> 	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
> 	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
> 	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
> 	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
> 	at io.quarkus.quartz.runtime.QuartzScheduler_Bean.get(Unknown Source)
> 	at io.quarkus.quartz.runtime.QuartzScheduler_Bean.get(Unknown Source)
> 	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:440)
> 	at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:453)
> 	at io.quarkus.arc.impl.ArcContainerImpl.instanceHandle(ArcContainerImpl.java:424)
> 	at io.quarkus.arc.impl.ArcContainerImpl.instance(ArcContainerImpl.java:233)
> 	at io.quarkus.arc.runtime.BeanLookupSupplier.get(BeanLookupSupplier.java:29)
> 	at io.quarkus.vertx.http.runtime.devmode.DevConsoleRecorder.addInfo(DevConsoleRecorder.java:41)
> 	at io.quarkus.deployment.steps.DevConsoleProcessor$runtimeTemplates-308161071.deploy_0(Unknown Source)
> 	at io.quarkus.deployment.steps.DevConsoleProcessor$runtimeTemplates-308161071.deploy(Unknown Source)
> 	... 13 more
> Caused by: org.quartz.ObjectAlreadyExistsException: Unable to store Trigger with name: 'executeCalculationJob_trigger' and group: 'io.quarkus.scheduler.Scheduler', because one already exists with this identification.
> 	at org.quartz.impl.jdbcjobstore.JobStoreSupport.storeTrigger(JobStoreSupport.java:1184)
> 	at org.quartz.impl.jdbcjobstore.JobStoreSupport$2.executeVoid(JobStoreSupport.java:1068)
> 	at org.quartz.impl.jdbcjobstore.JobStoreSupport$VoidTransactionCallback.execute(JobStoreSupport.java:3780)
> 	at org.quartz.impl.jdbcjobstore.JobStoreSupport$VoidTransactionCallback.execute(JobStoreSupport.java:3778)
> 	at org.quartz.impl.jdbcjobstore.JobStoreCMT.executeInLock(JobStoreCMT.java:245)
> 	at org.quartz.impl.jdbcjobstore.JobStoreSupport.storeJobAndTrigger(JobStoreSupport.java:1063)
> 	at org.quartz.core.QuartzScheduler.scheduleJob(QuartzScheduler.java:855)
> 	at org.quartz.impl.StdScheduler.scheduleJob(StdScheduler.java:249)
> 	at io.quarkus.quartz.runtime.QuartzScheduler.<init>(QuartzScheduler.java:200)
> 	... 32 more
> ```

I am pretty sure that the `identity` suggestion works and I've tested it before. The exception you are seeing is rather strange but possibly a different issue worth investigating. Do you mind preparing a reproducer for it and create an issue for it? Thanks

 

> I think I will:
> 
> 1. change the identity of all Quarkus 1.x schedulers to something different. That way new schedulers will be created as Qaurkus 2.x schedulers
> 2. deploy to production (cluster of services)
> 3. update the cron as needed and run flyway script to delete the old ones
> 
> Do you have another idea/suggestion? 

The other idea was to recreate the quartz table via a database migration using Flyway/Liqubase extensions, since at this point I suspect the corresponding tables might have been corrupted if you did manual modifications.

--- a/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
+++ b/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
@@ -37,6 +37,7 @@ import org.quartz.SchedulerFactory;
 import org.quartz.SimpleScheduleBuilder;
 import org.quartz.Trigger.TriggerState;
 import org.quartz.TriggerBuilder;
+import org.quartz.TriggerKey;
 import org.quartz.impl.StdSchedulerFactory;
 import org.quartz.simpl.InitThreadContextClassLoadHelper;
 import org.quartz.simpl.SimpleJobFactory;
@@ -179,8 +180,10 @@ public class QuartzScheduler implements Scheduler {
                             throw new IllegalArgumentException("Invalid schedule configuration: " + scheduled);
                         }
 
+                        JobDetail jobDetail = jobBuilder.build();
                         TriggerBuilder<?> triggerBuilder = TriggerBuilder.newTrigger()
                                 .withIdentity(identity, Scheduler.class.getName())
+                                .forJob(jobDetail)
                                 .withSchedule(scheduleBuilder);
 
                         Long millisToAdd = null;
@@ -194,18 +197,30 @@ public class QuartzScheduler implements Scheduler {
                                     .plusMillis(millisToAdd).toEpochMilli()));
                         }
 
-                        JobDetail job = jobBuilder.build();
                         org.quartz.Trigger trigger = triggerBuilder.build();
-                        if (!scheduler.checkExists(job.getKey())) {
-                            scheduler.scheduleJob(job, trigger);
-                            LOGGER.debugf("Scheduled business method %s with config %s", method.getMethodDescription(),
-                                    scheduled);
-                        } else {
-                            org.quartz.Trigger oldTrigger = scheduler.getTrigger(trigger.getKey());
+                        org.quartz.Trigger oldTrigger = scheduler.getTrigger(trigger.getKey());
+                        if (oldTrigger != null) {
                             scheduler.rescheduleJob(trigger.getKey(),
                                     triggerBuilder.startAt(oldTrigger.getNextFireTime()).build());
                             LOGGER.debugf("Rescheduled business method %s with config %s", method.getMethodDescription(),
                                     scheduled);
+                        } else if (!scheduler.checkExists(jobDetail.getKey())) {
+                            scheduler.scheduleJob(jobDetail, trigger);
+                            LOGGER.debugf("Scheduled business method %s with config %s", method.getMethodDescription(),
+                                    scheduled);
+                        } else {
+                            // TODO remove this code in 3.0, it is only here to ensure migration after the removal of
+                            // "_trigger" suffix and the need to reschedule jobs due to configuration change between build
+                            // and deploy time
+                            oldTrigger = scheduler.getTrigger(new TriggerKey(identity + "_trigger", Scheduler.class.getName()));
+                            if (oldTrigger != null) {
+                                scheduler.deleteJob(jobDetail.getKey());
+                                scheduler.scheduleJob(jobDetail, triggerBuilder.startAt(oldTrigger.getNextFireTime()).build());
+                                LOGGER.debugf(
+                                        "Rescheduled business method %s with config %s due to Trigger '%s' record being renamed after removal of '_trigger' suffix",
+                                        method.getMethodDescription(),
+                                        scheduled, oldTrigger.getKey().getName());
+                            }
                         }
                     }
                 }

